File: leetcode.com/python/362_Design_Hit_Counter.py
from collections import defaultdict


class HitCounter(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.timestamps = []
        self.timestampHitMap = defaultdict(int)

    def hit(self, timestamp):
        """
        Record a hit.
        @param timestamp - The current timestamp (in seconds granularity).
        :type timestamp: int
        :rtype: None
        """
        if timestamp < 1:
            return
        if not self.timestamps or self.timestamps[-1] != timestamp:
            self.timestamps.append(timestamp)
        self.timestampHitMap[timestamp] += 1

    def getHits(self, timestamp):
        """
        Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity).
        :type timestamp: int
        :rtype: int
        """
        hitCount, startIdx = 0, 0
        for i in range(len(self.timestamps) - 1, -1, -1):
            startTime = self.timestamps[i]
            if timestamp - startTime < 300:
                hitCount += self.timestampHitMap[startTime]
                startIdx = i
            else:
                if startTime in self.timestampHitMap:
                    del self.timestampHitMap[startTime]
        self.timestamps = self.timestamps[startIdx:]
        return hitCount

# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/362_Design_Hit_Counter.swift
import Foundation

class HitCounter {

    var timestampHitMap: [Int: Int] = [:]
    var timestamps: [Int] = Array()
    
    /** Initialize your data structure here. */
    init() {
        
    }
    
    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    func hit(_ timestamp: Int) {
        if timestamp < 1 {
            return
        }
        if self.timestamps.count < 1 || self.timestamps.last != timestamp {
            self.timestamps.append(timestamp)
        }
        self.timestampHitMap[timestamp, default: 0] += 1
    }
    
    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    func getHits(_ timestamp: Int) -> Int {
        var hitCount = 0
        var startIdx = 0
        for i in stride(from: self.timestamps.count - 1, to: -1, by: -1) {
            let startTime = self.timestamps[i]
            if timestamp - startTime < 300 {
                hitCount += self.timestampHitMap[startTime] ?? 0
                startIdx = i
            } else {
                if self.timestampHitMap[startTime] != nil {
                    self.timestampHitMap.removeValue(forKey: startTime)
                }
            }
        }
        self.timestamps = Array(self.timestamps[startIdx...])
        return hitCount
    }
}

/**
 * Your HitCounter object will be instantiated and called as such:
 * let obj = HitCounter()
 * obj.hit(timestamp)
 * let ret_2: Int = obj.getHits(timestamp)
 */
